components/file_system/src/iosnoop/biosnoop.rs

```diff
@@ -146,52 +145,19 @@ pub fn get_io_type() -> IOType {
     unsafe { *IDX.with(|idx| IO_TYPE_ARRAY[idx.0]) }
 }
 
-unsafe fn get_io_stats() -> Option<HashMap<IOType, IOStats>> {
-    if let Some(ctx) = BPF_CONTEXT.as_mut() {
-        let mut map = HashMap::default();
-        for e in ctx.stats_table.iter() {
-            let io_type = ptr::read(e.key.as_ptr() as *const IOType);
-            let read = std::intrinsics::atomic_load(e.value.as_ptr() as *const u64);
-            let write = std::intrinsics::atomic_load((e.value.as_ptr() as *const u64).add(1));
-            map.insert(io_type, IOStats { read, write });
-        }
-        Some(map)
-    } else {
-        None
-    }
-}
-
-pub struct IOContext {
-    io_stats_map: Option<HashMap<IOType, IOStats>>,
-}
-
-impl IOContext {
-    pub fn new() -> Self {
-        IOContext {
-            io_stats_map: unsafe { get_io_stats() },
-        }
-    }
-
-    pub fn delta_and_refresh(&mut self) -> HashMap<IOType, IOStats> {
-        if self.io_stats_map.is_some() {
-            if let Some(map) = unsafe { get_io_stats() } {
-                for (io_type, stats) in &map {
-                    self.io_stats_map
-                        .as_mut()
-                        .unwrap()
-                        .entry(*io_type)
-                        .and_modify(|e| {
-                            e.read = stats.read - e.read;
-                            e.write = stats.write - e.write;
-                        })
-                        .or_insert(stats.clone());
-                }
-
-                return self.io_stats_map.replace(map).unwrap();
+pub(crate) fn fetch_io_bytes(io_type: IOType) -> IOBytes {
+    unsafe {
+        if let Some(ctx) = BPF_CONTEXT.as_mut() {
+            let mut buf = [0 as u8; std::mem::size_of::<IOType>()];
+            std::ptr::write(buf.as_mut_ptr() as *mut IOType, io_type);
```

Comment:
Prompted by @yiwu-arbug's question I thought a bit about the unsafe code here, and these two lines don't look sound to me. If size_of<IOType> != 1 then I don't think there's any guarantee that this buffer is correctly aligned, which it needs to be for the cast back to `*mut IOType`. That is, I don't _think_ Rust has any guarantee that on-stack u8 arrays will be aligned to anything other than a u8.

It _looks_ to me that IOType is probably a u32 since it as a `repr(C)` enum.

You might test this with an assert that size of IOType == 1.

In general, it is not sound to cast u8 buffers into pointers of types that require larger than u8 alignment. If you do then you must _know_ that buffer is correctly aligned.

As written this type of code will often work most of the time, while still introducing undefined behavior, and could fail in rare scenarios.

The point of this code looks to be to pass an `&mut [u8]` that represents the given `IOType` to `ctx.stats_table.get`.

(Aside: it isn't clear to me why this buffer needs to be mutable, that is - why `bcc::Table::get` takes a mutable slice, as the only thing _in_ this buffer is a temporary C-style enum).

Instead of creating a u8 buffer and byte-copying the `io_type` into it, I think this code should instead be taking the approach of casting the given IOType to a buffer of u8.

You might do this by

- making io_type mutable
- casting a mutable pointer to `*mut u8`
- constructing a mutable u8 slice with `slice::from_raw_parts_mut(u8_ptr, size_of::<IOType>)`